File: 03day/part2.py


def solve(lines):
    a, b, c = map(set, lines)
    same, = a & b & c
    if ord('a') <= ord(same) <= ord('z'):
        return (ord(same) - ord('a')) + 1
    else:
        return (ord(same) - ord('A')) + 27


if __name__ == "__main__":
    n = 3
    lines = []
    with open("input.txt") as f:
        for line in f:
            lines.append(line.strip())
    by_three = [lines[i:i+n] for i in range(0, len(lines), n)]
    print(sum(map(solve, by_three)))

File: 03day/part2.hs
import Data.Char (isAsciiLower, ord)
import Data.List (foldr1, intersect, unfoldr)
import Data.Set (empty, fromList, intersection, toList, union)
import System.IO (IOMode (ReadMode), hClose, hGetContents, openFile)

chunk :: Int -> [a] -> [[a]]
chunk n = takeWhile (not . null) . unfoldr (Just . splitAt n)

score :: Char -> Int
score c
  | isAsciiLower c = (ord c - ord 'a') + 1
  | otherwise = (ord c - ord 'A') + 27

solve :: [String] -> Int
solve = score . head . toList . foldr1 intersection . map fromList

solution :: String -> Int
solution = sum . map solve . chunk 3 . lines

main :: IO ()
main = do
  handle <- openFile "input.txt" ReadMode
  content <- hGetContents handle
  print $ solution content
  hClose handle